### DISCUSSION TITLE
Branch Hygiene

### DISCUSSION THREAD

**Discussion by nyurik** [maplibre/maplibre#5] [General]
Currently we have a bit of a mess with branches, e.g. [this](https://github.com/maplibre/maplibre-gl-js/branches). We should only create git branches in the MapLibre repositories if they are dedicated to a specific goal, e.g. maintaining a released version (i.e. `v1.x`), or to collaboratively work on a major new feature (e.g. switching to typescript) before it is merged to main.

All other work should be done on contributor's own forks, and submitted via pull requests.

Without this discipline, we will end up with thousands of unmaintained branches.  Also we confuse our users because when they fork, they copy all those branches as well.

Perhaps we should restrict branch creation using github settings somehow?

cc: @HarelM @wipfli @astridx

**Comment by wipfli** [comment]
Sounds good to me. Thanks for sharing your concern.

**Reply by HarelM** [reply]
I don't share this concern yet... Only maintainers can create a branches so we can easily control this and make sure we remove branches according to some rules...
I agree we should make sure we clean up, but I do not agree we should force to fork for maintainers. This is only my personal perspective...

**Reply by nyurik** [reply]
@HarelM when you create a branch like `526-remove-non-typed-modules`, everyone who clones or forks the repo will get that branch. Moreover, when you delete it, they will keep it, and will need to clean it up.  That branch has no value for them -- it is purely noise. That's why this is usually a rule for many projects -- all contributors should have their own forks, and commit using pull requests.

P.S. also this is good from a personal perspective -- we separate what is agreed to by the community from what is a contribution by one of the people participating in the project - regardless if that person is a maintainer or a casual visitor.

P.P.S. This also has some impact on the community -- let's try not to minimize the separation between different levels of contributors -- when everyone contributes using the same way, everyone feels like they own the project semi-equally. So a casual contributor could gradually contribute more, but will not have to change to a different workflow of contributing later on, etc.  (granted that merge has to happen by maintainers)

**Reply by HarelM** [reply]
While I understand your point I do not agree with it. From my personal point of view as a maintainer forking and working on two different repos on my dev env is less efficient and less effective. I prefer to have velocity over the noise I may cause someone who fork this repo and may or may not contribute back to it.
It might be selfish, but I prefer practical over prude in most cases... :-)

**Reply by nyurik** [reply]
@HarelM I do agree that in some cases practicality wins over prude-ness. In this specific case, the cost of you pushing a change to a fork has near-zero cost, but allows the rest of the team, as well as more casual contributors to work with cleaner code and have fewer surprises. Also, it avoids the "us vs them" mentality -- all contributes in the same way. Lastly, it ensures that we all use the same tooling setup and problems are spotted early on - so if your contribution doesn't work with github workflows, it means it won't work for others, so it will get fixed better.

Same as for the coding style -- It is easier to code in one's own style, but the project establishes coding conventions to make sure we can all collaborate.  I think this is a small enough cost to pay to respect other contributors, especially when they may become maintainers in the future.

**Reply by HarelM** [reply]
Keeping my fork up to date is not near zero cost for me, and this is where we don't see eye to eye...
As I said I understand your points and agree with them in general, but your assumptions are not accurate for my use case, specifically.

**Reply by nyurik** [reply]
@HarelM I'll be more than happy to walk you through the git command to make it near zero for you, or at least get a better understanding of why it would cause any issues.  Let's sync on slack to find a good time for a video call

**Reply by Joxit** [reply]
Now with Github you can sync your fork via the UI 
![image](https://user-images.githubusercontent.com/5153882/142740081-00f3ffd0-6067-4107-b5c0-fad6a72b1b4b.png)

**Comment by astridx** [comment]
> Perhaps we should restrict branch creation using github settings somehow?

It may be a habit, but I know it from another project that everyone works in their own repo and that only PR is used to submit new things. I like this additional step because you can check once again whether everything is correct. It has happened to me twice now that I have accidentally created a branch in the main repository. That's why I would support this restriction.

**Reply by nyurik** [reply]
Thanks @astridx , can your branches listed in https://github.com/maplibre/maplibre-gl-js/branches be deleted?

**Reply by astridx** [reply]
Yes, those were the ones where I accidentally entered the maplibre remote instead of my own. I have just deleted them. Is this OK so?
